pkg/operator/vspherecontroller/vspherecontroller.go

```diff
@@ -456,32 +465,44 @@ func (c *VSphereController) createVCenterConnection(ctx context.Context, infra *
 	if !ok {
 		return fmt.Errorf("cloud config %s/%s does not contain key %q", cloudConfigNamespace, cloudConfig.Name, cloudConfig.Key)
 	}
-	cfg := new(vsphere.VSphereConfig)
-	err = gcfg.ReadStringInto(cfg, cfgString)
+	//cfg := new(vsphere.VSphereConfig)
+	//err = gcfg.ReadStringInto(cfg, cfgString)
+
+	// If we use infra to iterate through vcenters, do we need to load config?
+	cfg, err := vsphere.ReadConfig([]byte(cfgString))
 	if err != nil {
 		return err
 	}
 
 	// We no longer use the ConfigMap to store the vSphere config, so make sure to delete it
 	c.deleteConfigMapIfExists(ctx, legacyConfigMapName, c.targetNamespace)
 
-	secret, err := c.secretLister.Secrets(c.targetNamespace).Get(cloudCredSecretName)
-	if err != nil {
-		return err
-	}
-	userKey := cfg.Workspace.VCenterIP + "." + "username"
-	username, ok := secret.Data[userKey]
-	if !ok {
-		return fmt.Errorf("error parsing secret %q: key %q not found", cloudCredSecretName, userKey)
-	}
-	passwordKey := cfg.Workspace.VCenterIP + "." + "password"
-	password, ok := secret.Data[passwordKey]
-	if !ok {
-		return fmt.Errorf("error parsing secret %q: key %q not found", cloudCredSecretName, passwordKey)
-	}
+	for _, vcenter := range infra.Spec.PlatformSpec.VSphere.VCenters {
+		secret, err := c.secretLister.Secrets(c.targetNamespace).Get(cloudCredSecretName)
+		if err != nil {
+			return err
+		}
+		userKey := vcenter.Server + "." + "username"
+		username, ok := secret.Data[userKey]
+		if !ok {
+			return fmt.Errorf("error parsing secret %q: key %q not found", cloudCredSecretName, userKey)
+		}
+		passwordKey := vcenter.Server + "." + "password"
+		password, ok := secret.Data[passwordKey]
+		if !ok {
+			return fmt.Errorf("error parsing secret %q: key %q not found", cloudCredSecretName, passwordKey)
+		}
+
+		// just a hack for other function compatibility.
+		/*cfg := new(legacy.VSphereConfig)
+		cfg.Workspace.VCenterIP = vcenter.Server
+		cfg.Workspace.Datacenter = vcenter.Datacenters[0]
+		cfg.Workspace.DefaultDatastore = infra.Spec.PlatformSpec.VSphere.FailureDomains[0].Topology.Datastore
+		cfg.Global.InsecureFlag = true*/
```

Comment:
can we drop this commented code or are we leaving it in for historical reasons?

Reply:
These will be cleaned up shortly.  Still working on merging conflicts and fixing.